project/src/com/google/daggerqueryui/scripts/binding_graph.js

```diff
@@ -0,0 +1,212 @@
+   * @param {string} node a full name of the node
+   */
+  function getSimpleName(node) {
+    const firstIndex = node.indexOf('[');
+
+    if (firstIndex === -1) {
+      const components = node.split('.');
+      return components[components.length - 1];
+    }
+
+    const container = node.substring(0, firstIndex).split('.').slice(-1);
+    const content = getSimpleName(node.substring(firstIndex + 1, node.lastIndexOf(']')));
+    return `${container}[${content}]`;
+  }
+
+  /**
+   * Associates given node with an identifier, if the node was not represented in the graph.
+   *
+   * @param {string} node a node to be added if it doesn't exist
+   * @return {number} an identifier of a given node
+   */
+  function addNode(node) {
+    if (!nodesToNumbers.has(node)) {
+      nodesToNumbers.set(node, nodesToNumbers.size);
+    }
+
+    const id = nodesToNumbers.get(node);
+    if (nodes.get(id) === null) {
+      // Since angle brackets have a special meaning in html, we replace them with square brackets.
+      const fullName = node.replace(/</g, "[").replace(/>/g, "]");
+      const simpleName = getSimpleName(fullName);
+      nodes.add({id: id, label: simpleName, title: fullName});
+    }
+    return id;
+  }
+
+  /**
+   * Tries to remove the given node from a subgraph.
+   *
+   * <p>The operation is considered successful only if the given node does not connect
+   * to any other node in the graph.
+   *
+   * @param {number} nodeId an identifier of a node to be removed
+   */
+  function tryToRemoveNode(nodeId) {
+    if (getAllEdges(nodeId).length === 0) {
+      nodes.remove(nodeId);
+    }
+  }
+
+  /**
+   * Retrieves all edges where the start or destination nodes are equal to the given one.
+   *
+   * @param {number} nodeId an identifier of the given node
+   */
+  function getAllEdges(nodeId) {
+    return edges.get({
+      filter: function (edge) {
+        return (edge.from === nodeId || edge.to === nodeId);
+      }
+    });
+  }
+
+  /**
+   * Retrieves all edges where the start node is the specified one.
+   *
+   * @param {number} nodeId an identifier of the given node
+   */
+  function getAllDeps(nodeId) {
+    return edges.get({
+      filter: function (edge) {
+        return (edge.from === nodeId);
+      }
+    });
+  }
+
+  /**
+   * Removes an edge from a subgraph, and if the destination node
+   * becomes isolated, then removes it as well.
+   *
+   * @param edge an edge which will be removed
+   */
+  function removeEdge(edge) {
+    edges.remove(edge);
+
+    tryToRemoveNode(edge.to);
+  }
+
+  return {
+    /**
+     * Adds an edge from the <b>source</b> node to the <b>target</b>.
+     *
+     * @param {string} source the first node of the given edge
+     * @param {string} target the second node of the given edge
+     */
+    addEdge: function (source, target) {
+      const sourceId = addNode(source);
+      const targetId = addNode(target);
+
+      edges.add({from: sourceId, to: targetId});
+    },
+
+    /**
```

Comment:
It looks like this method isn't currently in use.
Is this functionality related to rdeps?

Reply:
Yes, it will later be used to remove previously displayed deps. Since this PR contains quite a few changes, let's remove this feature, I will add it to a PR that supports click-to-remove 🐝